aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -935,6 +951,79 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
   return std::tuple<Tensor&, Tensor&>(vals, vecs);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ eig ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+template <typename scalar_t>
+static void apply_eig(Tensor& vals_, Tensor& vecs_, const Tensor& self, bool eigenvectors) {
+#ifndef USE_LAPACK
+  AT_ERROR("symeig: LAPACK library not found in compilation");
+#else
+  int n = self.size(0);
+  Tensor wi = at::empty({n}, self.dtype());
+  Tensor wr = at::empty({n}, self.dtype());
+  char jobvr = eigenvectors ? 'V' : 'N';
+  int lda = n;
+  int info;
+  if (n > 0) {
+    scalar_t wkopt;
+    scalar_t* vecs_data = eigenvectors ? vecs_.data<scalar_t>() : nullptr;
+    int ldvr = eigenvectors ? n : 1;
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data<scalar_t>(), lda, wr.data<scalar_t>(), wi.data<scalar_t>(),
+      nullptr, 1, vecs_data, ldvr, &wkopt, -1, &info);
+
+    int lwork = (int)wkopt;
+    Tensor work = at::empty({lwork}, self.dtype());
+
+    lapackEig<scalar_t>('N', jobvr, n, self.data<scalar_t>(), lda, wr.data<scalar_t>(), wi.data<scalar_t>(),
+      nullptr, 1, vecs_data, ldvr, work.data<scalar_t>(), lwork, &info);
+  }
+
+  Tensor wr_ = at::slice(vals_, 1, 0, 1).squeeze(1);
+  Tensor wi_ = at::slice(vals_, 1, 1, 2).squeeze(1);
+  wr_.copy_(wr);
+  wi_.copy_(wi);
+#endif
+}
+
+static std::tuple<Tensor, Tensor> _eig_helper_cpu(const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dim() == 2, "A should be 2 dimensional");
+  TORCH_CHECK(self.size(0) == self.size(1), "A should be square");
+  TORCH_CHECK(self.isfinite().all().item<bool>(), "A should not contain infs or NaNs");
+  int n = self.size(0);
+  Tensor vecs_;
+  if (eigenvectors) {
+    vecs_ = at::empty({n, n}, self.dtype()).transpose(0, 1);
+  } else {
+    vecs_ = at::empty({0}, self.dtype());
+  }
+  Tensor vals_ = at::empty({n, 2}, self.dtype());
+  Tensor self_ = cloneBatchedColumnMajor(self);
+
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "eig_cpu", [&]{
+    apply_eig<scalar_t>(vals_, vecs_, self_, eigenvectors);
+  });
+  return std::tuple<Tensor&, Tensor&>(vals_, vecs_);
+}
+
+std::tuple<Tensor&, Tensor&> eig_out_cpu(Tensor& vals, Tensor& vecs, const Tensor& self, bool eigenvectors) {
+  TORCH_CHECK(self.dtype() == vals.dtype(), "dtype of eigenvalues tensor must match input");
+  TORCH_CHECK(!eigenvectors || (self.dtype() == vecs.dtype()), "dtype of eigenvectors tensor must match input");
+  Tensor vals_tmp, vecs_tmp;
+  std::tie(vals_tmp, vecs_tmp) = _eig_helper_cpu(self, eigenvectors);
+  vals.resize_as_(vals_tmp).copy_(vals_tmp);
+  vecs.resize_as_(vecs_tmp);
+  if (eigenvectors && vecs.is_contiguous()) {
+    vecs.transpose_(0, 1);
+  }
+  vecs.copy_(vecs_tmp);
```

Comment:
Can you please leave a comment why behavior for contiguous and non-contiguous vecs is different? It's very confusing. I also don't see tests for that, other than smoke tests for matching dtypes.

Reply:
Thanks for pointing that out, this code was not correct. The actual important thing is that the vecs output has to be column-major contiguous, or in other words, transposed-contiguous. The old TH code used `THTensor_(newTransposedContiguous)` to accomplish this. So I've updated this to properly resize to transposed-contiguous format.

There are already some asserts in `test_torch.py:test_eig` to check for this: `self.assertFalse(v.is_contiguous(), ...)`.

Reply:
Perhaps it would be better to update those asserts to be something like `self.assertTrue(v.transpose(0, 1).is_contiguous(), ...)`. I'll go ahead and make that change.

Reply:
> So if the code was not correct and tests were not catching it, clearly some tests are missing?

Yes, so checking `v.transpose(0, 1).is_contiguous()` will be more accurate.